### ISSUE TITLE
Button on Bluetooth headset does not answer or end Signal voice calls

### ISSUE THREAD

**Issue by cegg1** [signalapp/Signal-Android#7679]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Signal supports using Bluetooth headsets for voice calls. However, when an incoming Signal call rings and a bluetooth headset is paired and connected, pressing the multifunction button does not answer the Signal call. The call has to be answered manually from the screen, only then does the bluetooth set engage.

Even more surprisingly, pressing the multifunction button while a Signal call is in progress (and the BT set in use) has the effect of handing the call over to the regular speaker- rather than the expected effect of ending the call.

The same issue has been observed on Pixel 2 (stock 8.1) and on Nexus 5 (LineageOS 14.1).

### Steps to reproduce
- receive a Signal call, having a bluetooth headset paired and connected
- try to answer the call from the bluetooth headse

**Actual result:**  The call keeps ringing.
**Expected result:** The call is answered, similarly to regular phone call.

- while on a Signal call and speaking through a bluetooth headset
- try to hang up by pressing the button on the headset

**Actual result:**  The call gets handed over to the earpiece.
**Expected result:** The call is ended.

### Screenshots
Not relevant

### Device info
**Bluetooth headset:** Aukey EP-B40 Bluetooth 4.1
**Device:** Google Pixel 2
**Android version:** Stock Android 8.1.0 - OPM2.171019.029
**Signal version:** 4.18.3

### Link to debug log
Trying to pick up a Signal call by pressing the bluetooth button (nothing happens)
https://debuglogs.org/e98e666571435c9a13e6cac176ecdd4eda57ddd60a7127381c0c73625df697e0

Trying to end a Signal call by pressing the bluetooth button (call gets handed over to earpiece)
https://debuglogs.org/789eec7e8c8e1138cf86a1fe61ea803bcc1df35bbb80b4f21edc057eeb581b0a

**Comment by johanw666** [comment]
This is extremely complicated because Android reserves that functionality only for the official phone app. You can do that in a number of ways: the easiest way requires the MODIFY_PHONE_STATE permission, which is unfortunately only available for system apps. It would have the added advantage that an incomming "normal" call while you're on a Signal call gets busy.

I asked this question already on Stackoverflow once when it came up for my job but got no answers: https://stackoverflow.com/questions/47354818/how-to-make-a-voip-app-use-a-carkit-like-the-native-call-app

The only alternative that does not require root would be to implement the bluetooth communication on a low level yourself with BT sockets, but that is a lot of work, might not work correctly with all hardwareand require a lot of maintenance.

**Comment by johanw666** [comment]
Hangup is easy: you can detect the button being pressed, and when your call state is calling (the app knows when it is calling) it can hangup.

**Comment by tohojo** [comment]
I have also recently been bitten by this. A full integration would be awesome, of course, but just having hangup work would fix a lot of the annoyance; finding out that your call is still active minutes after you thought you hung up is a bit annoying :)

**Comment by johanw666** [comment]
I didn't work it out completely because the app I was working on didn't have this permission but the first set of results when I Google on `android accept calls MODIFY_PHONE_STATE` look promissing. And I believe that you didn't need system in very old Android versions so some old code could be usefull to check out too.

**Comment by vedaker** [comment]
I agree, I'm not sure there is an easy solution for this issue by using the existing Java APIs.
The reason is that the "Call" button event is most likely received by the low level Bluetooth driver but Android for some reason which I don;t understand is masking this event (security?).  So far I have not found a way to get the CALL events from the Java APIs.
You could however get a Bluetooth state change notification when you press on the "Call" button but this is not a good method to interpret the button press because the state might change by other events/actions in the system.

If you really must get the "Call" or "Disconnect" button events the solution might be to write a low level C code to parse the Bluetooth data stream, but this is most likely not an easy task to do.

**Comment by sellth** [comment]
This is only one of the issues related to Signal not using Android's proper call API, but handles calls as a media player instead. Related feature request: https://community.signalusers.org/t/implement-android-telecom-api-for-calls/41280

**Comment by cody-signal** [comment]
Try on 5.33 where we've integrated with the telecom system.

**Comment by sellth** [comment]
Sorry for the delay in getting back to you. Here are two new debug logs.
Device: Google Pixel 4a
Signal: 5.36.3

1. https://debuglogs.org/android/5.36.3/1bd8c900e5e44932cc9c613edab731c276ce19b4d7fbcc369699d51ff17fcc14
2. https://debuglogs.org/android/5.36.3/d50f3cc9d5e32ac0684f9ac77c8904899b641c6327a9307e401ea6bd9a4d1dbd

### First log:
- trying to answer call with Bluetooth earbuds button 
- sth. happens to the rington audio, but call is not answered
- answering call via the popover
- call audio actually playing out of the earphone speaker instead of Bluetooth, despite Signal claiming Bluetooth speaker being used
- switching back and forth between phone speaker and Bluetooth resolved this issue

### Second log:
- trying to answer call with earbuds button
- ringtone audio switches from earbuds to phone speaker
- answering call via the popover
- call audio actually playing out of the earphone speaker instead of Bluetooth, despite Signal claiming Bluetooth speaker is used 
- switching back and forth between phone speaker and Bluetooth resolved this issue

### After few more tries, this seems to be the behaviour:
- Bluetooth button disconnects ringtone audio sometimes, but always causes the subsequent call audio to not use the Bluetooth speaker (despite Signal claiming it does)
- when call is answered without pressing the Bluetooth button first, call audio uses the Bluetooth speaker just fine
- trying to hang up via the Bluetooth button does nothing
- there's also no "call bubble" created like for normal phone calls

**Comment by seung-pluto** [comment]
Phone: Google Pixel 6
Bluetooth Devices Tested: Jowave BH830, Google Pixel Buds Pro, Beats Studio Buds, Plantronics Voyager 5200
Android Version: 12 (July 2022 Patch Level)
Signal Versions: 5.43, 5.44

Wanted to write in to note that up until 5.44, I was having the exact same issue.  The [Telecom integration patched into 5.44](https://github.com/signalapp/Signal-Android/commit/e024541b8a746880c9ccade5180492352cd6f4c2) appears to have resolved this issue, at least for me.  I can confirm that now, the multifunction button on all devices tested will properly pick up on an incoming call and hang up when on a call.

**Comment by rajeevn1** [comment]
We still can't answer/end incoming calls on bluetooth headset, including car dashboards.  This seems to be a safety issue.  Looking at the code, it seems the allow list for telecom integration is empty.

Can you provide an update please.

**Comment by cody-signal** [comment]
We tried almost a half dozen times to get telecom integration working, but it's improperly implemented across multiple vendors and devices. It results in all kinds of really bad problems. At this time, there is no plan to return to it until we hear from Google that something has improved, which potentially could be their new telecom integration library but so far that has not been the case.

**Comment by rajeevn1** [comment]
Thanks for the update.  Hoping for resolution with:

https://android-developers.googleblog.com/2023/11/alpha-release-of-telecom-library.html

**Comment by Werkov** [comment]
Would the mentioned telecom library update apply to older Androids too?
Regardless of that, IIUC, the current implementation is not via telecom but "media player". Would it be simpler to implement the call ending functionality via this? (As that is what is troubling me much more than call answering.)

**Comment by comminutus** [comment]
Can this please be addressed? This is a pretty big limitation since often a headset is all I have and my phone has to be secured away. In this scenario I can't receive any signal calls.

**Comment by UserX404** [comment]
Just checked for the actual state of development. The lib is at [alpha 3](https://developer.android.com/jetpack/androidx/releases/core#core-telecom_version_10_2) since April. Seems there isn't much progress at this side.

**Comment by renaudallard** [comment]
The page specifies:
`androidx.core:core-telecom:1.0.0-alpha01 is released. This version is developed in an internal branch and should be used with Android 14 Beta 2.`
Doesn't that mean it could be "already" in android 14?

**Comment by UserX404** [comment]
Telecom API now reached RC1

**Comment by Viperized** [comment]
Since the Telecom API has been released for half a year, any news on this front? 

I am unable to pick up calls while driving a car. Current car says VOIP CALL like the call already started but it's just ringing. That's how I know it's signal. 

For the product as a whole, I think Bluetooth handling is a higher priority than Android Auto call UI. But that would be nice too since Whatsapp implemented it.

But basic Bluetooth handling is more universal for now.

Bumping.

**Comment by Omega3131** [comment]
I am having this same issue on a Pixel 8 running GrapheneOS. Cars, headsets, and other accesories would greatly benefit from this. I have my grandma on Signal for video calls with the family, and she also has an Xlink device to bridge her cell phone to her old landlines. Being able to voice call her over Signal and have her phones ring would be ideal.

If someone can direct me to an official bounty program I am willing to put up $150 to aid development of this fix.

**Comment by Flatmotion1** [comment]
Same issue for years now.
Asus zenfone 10.

I ride motorcycle and the ability to pick up calls via the headset is a necessity. I can't use signal while riding a bike, accept or hang up calls.

**Comment by alex-signal** [comment]
Hi guys, just a heads up, we are looking into the telecom api issues, and we've found bugs in the api itself that we've reported and are tracking.

**Comment by Flexmaen** [comment]
I can confirm the issue with a Logilink BT0059.
Phone keeps ringing, ringing will be interrupted for a moment, and then goes on...
